[SPARK-32381][CORE][SQL] Move and refactor parallel listing & non-location sensitive listing to core

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This moves and refactors the parallel listing utilities from `InMemoryFileIndex` to Spark core so it can be reused by modules beside SQL. Along the process this also did some cleanups/refactorings:

- Created a `HadoopFSUtils` class under core
- Moved `InMemoryFileIndex.bulkListLeafFiles` into `HadoopFSUtils.parallelListLeafFiles`. It now depends on a `SparkContext` instead of `SparkSession` in SQL. Also added a few parameters which used to be read from `SparkSession.conf`: `ignoreMissingFiles`, `ignoreLocality`, `parallelismThreshold`, `parallelismMax ` and `filterFun` (for additional filtering support but we may be able to merge this with `filter` parameter in future).
- Moved `InMemoryFileIndex.listLeafFiles` into `HadoopFSUtils.listLeafFiles` with similar changes above.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Currently the locality-aware parallel listing mechanism only applies to `InMemoryFileIndex`. By moving this to core, we can potentially reuse the same mechanism for other code paths as well. 

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Since this is mostly a refactoring, it relies on existing unit tests such as those for `InMemoryFileIndex`.

holdenk committed:
Start working on re-implementing the resolution logic to avoid getting locations

holdenk committed:
Revert "Start working on re-implementing the resolution logic to avoid getting locations"

This reverts commit bbe63444b4cac0bd6858f1c684a38578e9c4ab5e.

holdenk committed:
Start moving InMemoryFileIndex over to HadoopFSUtils

holdenk committed:
Get the SQL layer compiling against the common shim layer

holdenk committed:
Keep working on plumbing through the type info we need to avoid directly manipulating bytecode

holdenk committed:
Ok core compiles now

holdenk committed:
Get the input filter based on the jobContext

holdenk committed:
Backout some small changes we don't need anymore

holdenk committed:
Revert the class change to BinaryFileRDD we don't depend on that anymore.

holdenk committed:
Fix dropped annotation

holdenk committed:
Back out NewHadoopRDD changes, we'll expose a trait instead and have people mix it in

holdenk committed:
Rework the HadoopFSUtils to: use the serilizablility of block locations, try and avoid expensive S3A location lookups (and other general file systems), etc.

holdenk committed:
Fix the bug and remove default params so it's more difficult to write

holdenk committed:
Remove un-used changes to NewHadoopRDD

holdenk committed:
Put the utility function in HadoopFSUtils to make it easier.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

sunchao committed:
Fix NPE

sunchao committed:
Defaults ignoreLocality to false and small cleanups

sunchao committed:
Don't use null as BlockLocations

sunchao committed:
Revert changes on WholeTextInputFileFormat

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented:
cc @holdenk

SparkQA commented:
**[Test build #127705 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127705/testReport)** for PR 29471 at commit [`5529047`](https://github.com/apache/spark/commit/5529047f44b0f451f63e79651a19704721d21920).

viirya commented:
btw, we should better use Spark's PR template. But this is still in WIP, I suppose it will fill the template later.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32329/
Test PASSed.

holdenk commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+    filterFun: Option[String => Boolean] = None): Seq[(Path, Seq[FileStatus])] = {
+    HiveCatalogMetrics.incrementParallelListingJobCount(1)
+
+    val serializableConfiguration = new SerializableConfiguration(hadoopConf)
```
I remember in the other PR it was mentioned this is now serializable on its own, is that true with the versions of Hadoop we use/can we avoid using this wrapper?

sunchao replied:
Ah - let me quickly check on this

sunchao replied:
So no - it still doesn't inherit from `Serializable`. No change has happened to the signature since 2009.

holdenk replied:
Thanks for checking :)

holdenk reviewed (commented):
Thanks for picking up this PR. Let me know when you want me to do a review pass (I see it's still marked as draft).

SparkQA commented:
**[Test build #127705 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127705/testReport)** for PR 29471 at commit [`5529047`](https://github.com/apache/spark/commit/5529047f44b0f451f63e79651a19704721d21920).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127705/
Test PASSed.

sunchao committed:
Remove unused import

SparkQA commented:
**[Test build #127709 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127709/testReport)** for PR 29471 at commit [`6f5c7e5`](https://github.com/apache/spark/commit/6f5c7e5dd52e484ec56e1cfd8435d4f00f682c57).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32333/
Test PASSed.

sunchao commented:
Thanks @viirya - fixed the PR description.

@holdenk this is ready for review now. I removed the changes on `WholeTextFileInputFormat` to make this focusing on refactoring (i.e., moving the code from SQL to core). We can do follow-ups if later on we decide to swap the listing logic for Hadoop file input formats as well. 

Other than that, a few fixes from the version I inherited from you:
- fixed a NPE when `BlockLocations` is null - I think we should not convert `FileStatus` to `LocatedFileStatus` and initialize block locations with null. 
- changed `ignoreDataLocality` default to false, to match that in SQL
- removed extra handling of `FileNotFoundException` which seems not necessary since the default behavior of `FileSystem` won't throw the exception, as mentioned offline.

SparkQA commented:
**[Test build #127709 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127709/testReport)** for PR 29471 at commit [`6f5c7e5`](https://github.com/apache/spark/commit/6f5c7e5dd52e484ec56e1cfd8435d4f00f682c57).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127709/
Test PASSed.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -443,6 +443,40 @@ package object config {
+        "of detected paths exceeds this value during partition discovery, it tries to list the " +
+        "files with another Spark distributed job. This configuration is effective only when " +
+        "using file-based using HadoopFSUtils")
+      .version("1.5.0")
```
This configuration looks new. Not in 1.5.0.

HyukjinKwon replied:
BTW, where are these configurations used?

sunchao replied:
Oops will fix. Yes these are new core configurations that copied from [SQLConf](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L990). I don't have any better idea on how to solve the duplication yet.

HyukjinKwon commented:
cc @steveloughran FYI from https://github.com/apache/spark/pull/29179

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+   *
+   * @return for each input path, the set of discovered files for the path
+   */
+  @DeveloperApi
```
I guess we don't have to annotate each method since the top object is already annotated. Looks like we should add version info as well.

sunchao replied:
OK will do.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+              maxParallelism = 0)
+            (path, leafFiles)
+          }.iterator
+        }.collect() // TODO(holden): should we use local itr here?
```
Can you file a JIRA for a `TODO`? I think that's what people usually do for trackability.

HyukjinKwon replied:
If it's planned to be addressed within this PR, feel free to turn this PR to a draft.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala:
```diff
@@ -290,9 +290,9 @@ class FileIndexSuite extends SharedSparkSession {
           }
         }
         HiveCatalogMetrics.reset()
```
I think we can remove all changes here. There's no strong preference between `==` and `===` when it compares a scalar value. Sometimes `==` provides a better assert message in this case.

sunchao replied:
Will remove these.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+    val statusMap = try {
+      val description = paths.size match {
+        case 0 =>
+          s"Listing leaf files and directories 0 paths"
```
nit: looks no need to add the leading `s`.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+   * @return for each input path, the set of discovered files for the path
+   */
+  @DeveloperApi
+  def parallelListLeafFiles(sc: SparkContext, paths: Seq[Path], hadoopConf: Configuration,
```
If you target to expose this to Java users as well, we shouldn't use Scala friendly collections like `Seq` or `Option`.

zsxwing replied:
Since we need to pass `SparkContext` into this method, why not add it to SparkContext and return an RDD instead? Then the user can use RDD APIs to do other work in parallel, for example, add a filter to avoid pulling all paths to the driver.

sunchao replied:
Thanks. I'll address this in a separate PR. 

I'm not sure about adding this to RDD API though: the things done are very specific, such as handling `FileNotFoundException`, dealing with locality, threshold on whether to enable/disable parallelism etc. It is composed of basic API calls like `parallelize` and `mapPartitions`, and I'm not sure which part of this is general enough to be extracted as a separate RDD API.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+        case s =>
+          s"Listing leaf files and directories for $s paths:<br/>${paths(0)}, ..."
+      }
+      sc.setJobDescription(description) // TODO(holden): should we use jobgroup?
```
ditto, shell we file a JIRA?

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala:
```diff
@@ -147,32 +142,6 @@ class InMemoryFileIndex(
-      length: Long)
-
-  /** A serializable variant of HDFS's FileStatus. */
-  private case class SerializableFileStatus(
```
Looks like it's not just simply moving the codes but includes code refactoring. Can we separate PRs? one for refactoring and the other one for moving the codes.

This listing is very critical to Spark SQL and other APIs. I would like to be very cautious on touching it.

sunchao replied:
Yeah good suggestion. I'll limit this PR to only refactoring then. Should I create another JIRA for the cleanup PR? I'm pretty new to Spark :)

holdenk replied:
Yeah a seperate JIRA is best.

gengliangwang commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala:
```diff
@@ -516,6 +516,7 @@ class FileIndexSuite extends SharedSparkSession {
       SQLConf.get.setConf(StaticSQLConf.METADATA_CACHE_TTL_SECONDS, previousValue)
     }
   }
+
```
nit: unnecessary change

gengliangwang commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+   * @return for each input path, the set of discovered files for the path
+   */
+  @DeveloperApi
+  def parallelListLeafFiles(sc: SparkContext, paths: Seq[Path], hadoopConf: Configuration,
```
Nit: Please follow the style guide https://github.com/databricks/scala-style-guide#spacing-and-indentation to put each param in each line

gengliangwang commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+              filter = filter,
+              ignoreMissingFiles = ignoreMissingFiles,
+              ignoreLocality = ignoreLocality,
+              isSQLRootPath = areSQLRootPaths,
```
shall we have a consistent name here?

holdenk replied:
So the is/are is because listLeafFiles takes a single name and this method takes in a list of files.

gengliangwang commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -443,6 +443,40 @@ package object config {
       .booleanConf
       .createWithDefault(false)
 
+  val PARALLEL_PARTITION_DISCOVERY_THRESHOLD =
```
It seems that this config is not used

gengliangwang commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -443,6 +443,40 @@ package object config {
+      .intConf
+      .checkValue(parallel => parallel >= 0, "The maximum number of paths allowed for listing " +
+        "files at driver side must not be negative")
+      .createWithDefault(2)
```
The default values of `spark.sql.sources.parallelPartitionDiscovery.threshold` is 32, why is it 32 here?

viirya commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+  def parallelListLeafFiles(sc: SparkContext, paths: Seq[Path], hadoopConf: Configuration,
+    filter: PathFilter, areSQLRootPaths: Boolean, ignoreMissingFiles: Boolean,
+    ignoreLocality: Boolean, maxParallelism: Int,
+    filterFun: Option[String => Boolean] = None): Seq[(Path, Seq[FileStatus])] = {
```
Follow the above comment https://github.com/apache/spark/pull/29471/files#r474382279, I am not sure if `String => Boolean` is Java-friendly? If not consider to expose to Java users, seems okay.

sunchao replied:
Thanks. I'll limit the current PR to refactoring and address this in a later PR.

viirya commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+    filter: PathFilter, areSQLRootPaths: Boolean, ignoreMissingFiles: Boolean,
+    ignoreLocality: Boolean, maxParallelism: Int,
+    filterFun: Option[String => Boolean] = None): Seq[(Path, Seq[FileStatus])] = {
+    HiveCatalogMetrics.incrementParallelListingJobCount(1)
```
Put `HiveCatalogMetrics` here looks strange. This is a "Metrics for access to the hive external catalog". Should we skip this or create another metrics?

sunchao replied:
Good catch. It seems inappropriate here. Not sure what is the best way to plugin Hive/SQL metrics here but I'll think over it later.

holdenk replied:
We could pass in a callback for listing metrics?

viirya replied:
Or call this `HiveCatalogMetrics.incrementParallelListingJobCount(1)` in SQL side before calling `parallelListLeafFiles`?

holdenk replied:
So the problem with that is that if we have a root directory with many sub-directories in it, we may initially choose to do non-parallel listing and then as the sub directories build up switch to parallel listing.

sunchao replied:
Yeah move this to SQL will not work - I think we can perhaps add a callback later just like @holdenk suggested (if you don't mind leaving it here in this PR). 

On the other hand, I think the `HiveCatalogMetrics` is already misleading - `InMemoryFileIndex` is used by non-Hive data sources like file-based ones and this particular metric itself has nothing to do with Hive catalog. Perhaps a better approach is to move the metrics to some place in core that are for storage-specific things. In future we can include more such as listing total time etc.

viirya replied:
A callback sounds good. We can do it later.

holdenk replied:
Sounds good to me, @sunchao want to file a JIRA for switching this to a callback?

sunchao replied:
Sure filed https://issues.apache.org/jira/browse/SPARK-32880 for the follow-ups.

zsxwing commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,259 @@
+   * Lists a collection of paths recursively. Picks the listing strategy adaptively depending
+   * on the number of paths to list.
+   *
+   * This may only be called on the driver.
```
nit: missing parameter docs. It's pretty important for an API that has 9 parameters.

sunchao committed:
Address comments and limit this to only refactoring

sunchao committed:
Remove Hive metrics call and nits

sunchao committed:
Make checkstyle happy

sunchao committed:
Fix checkstyle issue (for real) and change to Private API

sunchao committed:
Fix unidoc

sunchao committed:
Bring back parallel listing metrics

sunchao committed:
Fix style issue again (sigh)

steveloughran reviewed (commented):
I've got some opinions about how to list more efficiently, but given this is a movement of code, I'd propose saving that for a followup. Note that the InMemoryFileIndex does appear in flamegraphs when you start worrying about spark & cloud performance...

steveloughran commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,360 @@
+      fs match {
+        // DistributedFileSystem overrides listLocatedStatus to make 1 single call to namenode
+        // to retrieve the file status with the file block location. The reason to still fallback
+        // to listStatus is because the default implementation would potentially throw a
```
I would just switch to the incremental one *everywhere*

HDFS likes it because when you have many, many files in a path they can release the lock on the NN; for object stores they always have to page in (s3, azure, GCS)....which gives the implementors an opportunity to move from the default implementation to exposing the incremental one. Add this and we can just let the relevant teams know.

w.r.t CommonPathCapabilities, I suppose we could add one which declares that the listing ops are paged. But do you really want code to try and be that clever? I'm trying to use that feature to mark up

* where optional features will fail with UnimplementedException or similar without you having to try using them (append, truncate, xattrs)
* where there are fundamental semantics its important for algorithms to worry about (hflush)
* where your store wants to expose some state to callers (e.g. s3a exposing its HADOOP-13230 dir marker policy)

If you want a performance option, we could add one.

BTW, PathCapabilities is in hadoop-3.2.x now, will be in next release.  I might do it for 3.1 too...it makes a good way to programmatically/CLI probe for s3a dir marker policy, see.

Also, cloudstore has some CLI commands for the list calls (and pathcapabilities) to help explore what's going on, including, on s3a, listing cost in #of HTTP requests. Worth playing with to see what is good/bad, though as Mukund has been doing lots of work on 3.3.x s3a list, it will look worse than you'd expect on a dir treewalk,
https://github.com/steveloughran/cloudstore

holdenk replied:
I think changing this in a follow up PR sounds fine to me. I'd like to us to use the faster method by default and fall back on exception, but that could be a follow on. Want to file a JIRA for broadening the scope of using the faster method?

sunchao replied:
Thanks @steveloughran !

> I would just switch to the incremental one everywhere

Yes agree. This will help to simplify the code here a lot, although I guess we need to make sure this is enforced on the Hadoop side and that any future derived `FileSystem` implementations won't inherit the default behavior.

Back to the code, seems the main reason why it calls `listStatus` in the else case (e.g., not `DistributedFileSystem`) is because it want to save `getBlockLocations` cost by calling that later on filtered leaf files. We can potentially use `FilterFileSystem` here and pass the filter to it instead. I'll try to explore this in a follow-up PR.

> Want to file a JIRA for broadening the scope of using the faster method?

Yes will do that. The PR will be about improving and cleanup this utility class.

steveloughran replied:
w.r.t the faster (incremental) calls, yes, something to consider next. At the very least, you will be able to collect and report/aggregate stats. For example, here is me getting the stats on a large/deep list just by calling toString on the RemoteIterator after it's done its work

```
Listing statistics:
  counters=((object_list_request=1) (object_continue_list_request=40)); gauges=(); minimums=((object_list_request.min=1298) (object_continue_list_request.min=414)); maximums=((object_continue_list_request.max=746) (object_list_request.max=1298)); means=((object_continue_list_request.mean=(sum=18210, samples=39, mean=466.9231)) (object_list_request.mean=(sum=1298, samples=1, mean=1298.0000))); 

2020-08-27 13:25:21,122 [main] INFO  tools.MarkerTool (DurationInfo.java:close(98)) - marker scan s3a://landsat-pds/: duration 0:19.563s
Listed 40000 objects under s3a://landsat-pds/
```
Now, who wouldn't like to know things like that. And ideally, collect across threads and merge back in.

As an aside, looked at {{org.apache.hadoop.mapred.LocatedFileStatusFetcher}}. This does multithreaded status fetching and collects those stats. Although it's tagged Private, I've noticed Parquet uses it so my next PR will convert to public/evolving and document the fact. 

If you could use that, we could look @ evolving it better, especially returning a RemoteIterator of results which we could incrementally fill in across threads rather than block for the final results. Anything which makes it possible for app code to process data (read footers, etc) while the listing goes on has significant benefit in the World of Object Stores

sunchao replied:
Yes, stats will be nice. Looking forward to that 👍 (is it tracked by [this PR](https://github.com/apache/hadoop/pull/2069)?)

w.r.t `LocatedFileStatusFetcher`, yes it will be great if this is exposed as a public API. I think we should also consider moving this to another module such as hadoop-common, and perhaps adding an option to turn on/off locality (so that we don't have to get the block locations if they are not needed). Pairing it with batch listing can also potentially help (although this is only available in HDFS currently).

Instead of multiple threads, Spark currently use a distributed job so I'm not sure whether there will be any regression by doing that, but we can explore this later.

steveloughran replied:
> w.r.t LocatedFileStatusFetcher, yes it will be great if this is exposed as a public API. I think we should also consider moving this to another module such as hadoop-common, and perhaps adding an option to turn on/off locality (so that we don't have to get the block locations if they are not needed). Pairing it with batch listing can also potentially help (although this is only available in HDFS currently).

I'd worry about the effect of moving the package as is. It'd have to be a copy and paste, or move and subclass

steveloughran replied:
@sunchao -yes, that's the stats API. Now is the time to review it and tell me where it doesn't suit your (perceived future) needs. I've been playing with a Parquet branch which uses it, not in Spark itself.

sunchao replied:
Thanks for the info @steveloughran . I'll check out the PR (it is a very big one though).

holdenk commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,360 @@
+/**
+ * Utility functions to simplify and speed-up file listing.
+ */
+@Private
```
Private maybe seems a bit too restrictive in scope, what about DeveloperAPI?

sunchao replied:
@zsxwing suggested to make this private in the meanwhile and change it after the follow-up PR is done.

zsxwing replied:
Yep. I suggested to make this private. I think these APIs are not ready to expose yet. For example, 10 parameters  in a method is not user friendly. It's better to design a better API for this.

sunchao commented:
@HyukjinKwon @gengliangwang @viirya @zsxwing this PR is mostly a refactoring now. Could you take another look? Thanks!

zsxwing commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,360 @@
+ * Utility functions to simplify and speed-up file listing.
+ */
+@Private
+object HadoopFSUtils extends Logging {
```
Could you make this `private[spark]` instead? `@Private` is usually used in Java codes. Here is the doc for `@Private`

```
This should be used only when the standard Scala / Java means of protecting classes are
insufficient.  In particular, Java has no equivalent of private[spark], so we use this annotation
in its place.
```

sunchao replied:
Sure. Will do.

sunchao committed:
Switch to private[spark]

SparkQA commented:
**[Test build #127973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127973/testReport)** for PR 29471 at commit [`86c2013`](https://github.com/apache/spark/commit/86c2013abbf5b0365813b502cbfc9e2a5d5bc224).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32599/
Test PASSed.

SparkQA commented:
**[Test build #127973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127973/testReport)** for PR 29471 at commit [`86c2013`](https://github.com/apache/spark/commit/86c2013abbf5b0365813b502cbfc9e2a5d5bc224).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127973/
Test PASSed.

viirya commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,359 @@
+   * @param paths Input paths to list
+   * @param hadoopConf Hadoop configuration
+   * @param filter Path filter used to exclude leaf files from result
+   * @param areSQLRootPaths Whether the input paths are SQL root paths
```
Can we add more few words for `areSQLRootPaths`? Seeing SQL in `core` is already a bit strange,  so it's nicer to let developers can quickly get better idea just from reading doc.

sunchao replied:
Yes this is unfortunate. I think this parameter doesn't have to be visible to the callers though as it is set to true on the initial call and false on subsequent recursive calls. We can potentially add another overloaded method without this parameter and make this one private.

holdenk replied:
I like that idea @sunchao

viirya commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,359 @@
+   * @param parallelismThreshold The threshold to enable parallelism. If the number of input paths
+   *                             is smaller than this value, this will fallback to use
+   *                             sequential listing.
+   * @param parallelismMax The maximum parallelism for listing. If the number of input paths is
+   *                           larger than this value, parallelism will be throttled to this value
+   *                           to avoid generating too many tasks.
```
Indent looks wrong?

viirya commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,359 @@
+  // scalastyle:off argcount
+  /**
+   * Lists a single filesystem path recursively. If a `SparkContext`` object is specified, this
+   * function may launch Spark jobs to parallelize listing based on parallelismThreshold.
```
``` `parallelismThreshold` ```

viirya commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,359 @@
+
+  // scalastyle:off argcount
+  /**
+   * Lists a single filesystem path recursively. If a `SparkContext`` object is specified, this
```
A redundant `

sunchao committed:
Address comments

SparkQA commented:
**[Test build #128051 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128051/testReport)** for PR 29471 at commit [`bfa37cc`](https://github.com/apache/spark/commit/bfa37cc6e9e3b07e366cab982559d49f94a93248).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32677/
Test PASSed.

SparkQA commented:
**[Test build #128051 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128051/testReport)** for PR 29471 at commit [`bfa37cc`](https://github.com/apache/spark/commit/bfa37cc6e9e3b07e366cab982559d49f94a93248).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128051/
Test PASSed.

steveloughran commented:
BTW, wrote something up on listing.
https://github.com/steveloughran/engineering-proposals/blob/trunk/listing-performance.md

anywhere you do listStatus(path): List[FileStatus], switch to listStatusIterator, but, if the returned iterator is Closeable, make sure you close it after. Then I or a someone else will not only add the s3a and abfs speedups (alongside today's HDFS), I'll do the same for the local FS.

steveloughran commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,360 @@
+              }
+
+              SerializableFileStatus(
+                status.getPath.toString,
```
> since FileStatus is serializable, any subclass of it should also be, isn't it?

"should"; there's no tests for it, and nothing in the java language which stops you adding references to non-serializable objects. Exception has this exact issue

holdenk reviewed (approved):
LGTM, if no one objects I'll merge this on Monday so we can unblock the follow on work that folks seem interested in.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,360 @@
+      ignoreLocality: Boolean,
+      isRootPath: Boolean,
+      filterFun: Option[String => Boolean],
+      parallelismThreshold: Int,
```
This seems a new parameter that did not exist before. Why do we need this? If people want parallelized listing, people can invoke `parallelListLeafFiles` above.

sunchao replied:
This is because `listLeafFiles` also recursively call `parallelListLeafFiles` inside so we need a way to pass down the arguments. These used to be read from conf in `SparkSession` but now made explicit as parameters, as we no longer have a session object.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:
```diff
@@ -0,0 +1,360 @@
+        s"the following files were missing during file scan:\n  ${missingFiles.mkString("\n  ")}")
+    }
+
+    resolvedLeafStatuses.toSeq
```
Why do we need to call `toSeq`? Seems like we didn't do before.

sunchao replied:
You're right - I don't think this is needed. Will remove.

HyukjinKwon commented:
@sunchao Let's update PR title as well. This PR doesn't "expose" something but just moves the codes within the codebase. Also:
> Along the process this also did some cleanups/refactorings.

Do you mind clarify the additional diffs this PR introduces? I took a cursory look. Seems there are some diffs above.

HyukjinKwon commented:
Also, let's make sure file s JIRA at https://github.com/apache/spark/pull/29471#discussion_r486626261 as @holdenk suggested. I agree with @viirya's comment there.

sunchao commented:
Thanks @HyukjinKwon and @holdenk for the review! I updated the PR title as well as description. Also created [SPARK-32880](https://issues.apache.org/jira/browse/SPARK-32880) for the follow-up work.

sunchao committed:
Remove redundant .toSeq

SparkQA commented:
**[Test build #128666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128666/testReport)** for PR 29471 at commit [`2d8e64d`](https://github.com/apache/spark/commit/2d8e64d72a014ed5067be5772f8b2db9129ce2f2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33288/
Test PASSed.

SparkQA commented:
**[Test build #128666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128666/testReport)** for PR 29471 at commit [`2d8e64d`](https://github.com/apache/spark/commit/2d8e64d72a014ed5067be5772f8b2db9129ce2f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128666/
Test PASSed.

sunchao commented:
ping @HyukjinKwon @holdenk : anymore comments?

holdenk commented:
Looks like we've reached a lazy consesus here I'll merge this today :)

holdenk commented:
Ok I meant to merge this yesterday but I got distracted with the K8s stuff.

holdenk commented:
Merged to the current development branch :)

sunchao commented:
Thanks @holdenk for the merge, and all for the review comments!
